### PR TITLE
Zephyr v3.6.0 xt rpi5 dom0 wip

### PR DESCRIPTION
**Pull request by GrygiriiS**
First iteration of enabling RPI 5 in Zephyr and adding Dom0 functionality

### CODE REVIEW

**Comment by GrygiriiS** [comment]
@oleksiimoisieiev @dsemenets @xakep-amatop @Deedone

**Comment by oleksiimoisieiev** [comment]
lgtm 
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

**Comment by GrygiriiS** [comment]
Some checks failed on RPI5 initial patches as those patches ported as is from Zephyr ML. Those patches will need to be updated later any way when (if) they'll be merged in Zephyr ML. I have no intention to fix those CI warnings now.

UPD: fixed

File: boards/arm64/xenvm/xenvm_defconfig [hunk 1]
```diff
@@ -18,3 +18,4 @@ CONFIG_LOG_MODE_MINIMAL=n
 CONFIG_USERSPACE=n
 
 CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME=y
+CONFIG_XEN_REGIONS=n
```
Comment (firscity) [inline boards/arm64/xenvm/xenvm_defconfig:21]: It is better to add proper extended region, rather than disabling it completely. They are generated for every domain, you just need to add proper value to device tree. Also, such changes should not be located in board defconfig, it is better to move it to some test sample config you are using for building.

File: boards/arm64/rpi_5/Kconfig.board [hunk 2]
```diff
@@ -4,3 +4,7 @@
 config BOARD_RPI_5
 	bool "Broadcom BCM2712"
 	depends on SOC_BCM2712
+
+config BOARD_RPI_5_XEN_DOM0
```
Comment (firscity) [inline boards/arm64/rpi_5/Kconfig.board]: This whole commit is something improper. The basic idea of snippets - is extending or overwritting original board behavior: you use RPI5 and want to use Xen dom0 on it? Just apply `xen_dom0` overlay, no other changes or build targets should be needed. All Xen related actions should be done inside a snippet.

File: submanifests/zephyr-xenlib.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Copyright (C) 2024 EPAM Systems.
+#
+# Enable zephyr-xenlib support
+
+manifest:
```
Comment (firscity) [inline submanifests/zephyr-xenlib.yaml]: Please, provide proper commit message, I am out of context what is happening here and why we need this inside Zephyr

File: include/zephyr/xen/public/domctl.h [hunk 4]
```diff
@@ -20,7 +20,7 @@
 #include "grant_table.h"
 #include "memory.h"
 
-#define XEN_DOMCTL_INTERFACE_VERSION 0x00000015
+#define XEN_DOMCTL_INTERFACE_VERSION 0x00000016
```
Comment (firscity) [inline include/zephyr/xen/public/domctl.h:23]: Is it actually enough and no changes were made in these structures below? I have some doubts about it

File: boards/arm64/rpi_5/Kconfig.defconfig [hunk 5]
```diff
@@ -17,3 +17,10 @@ config XEN_DOM0
 	default y
 
 endif # BOARD_RPI_5_XEN_DOM0
+
+if BOARD_RPI_5_XEN_DOMD
```
Comment (firscity) [inline boards/arm64/rpi_5/Kconfig.defconfig]: This is also not okay to made separate board for this. You may use custom sample/application, please check other applications and how they override device tree and Kconfigs.

**Review by firscity — COMMENTED** [review]
Please, fix checkpatch warnings.

Also commits
1. `dts: arm64: broadcom: bcm2712: add psci and compatible` - add proper explanation for this
2. `drivers: xen: add XEN_EVENTS Kconfig option`
`but thay might be not used.` - > `but they might be not used.` 
3. `boards: arm64: xenvm: reduce hypervisor node memory` - please, mark this as a "WA" in commit topic, I will provide a proper solution soon.

File: boards/arm64/xenvm/xenvm_defconfig (hunk 1 above)
Comment (GrygiriiS) [reply boards/arm64/xenvm/xenvm_defconfig:21]: Help please - what should be the value for extended region? 

Right now the xenvm build is broken due to ext regions issue, so it has to be fixed within "xenvm" platform instead of adding a w/a fix in each sample/test.

File: boards/arm64/rpi_5/Kconfig.board (hunk 2 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/Kconfig.board]: I simply like this approach more as it allows to keep all platform data in one place, so added it as alternative to snippet. Np. will drop.

File: submanifests/zephyr-xenlib.yaml (hunk 3 above)
Comment (GrygiriiS) [reply submanifests/zephyr-xenlib.yaml]: It adds  zephyr-xenlib to this particular project. Not sure what to add here :(

https://docs.zephyrproject.org/latest/develop/modules.html#integrate-modules-in-zephyr-build-system

File: include/zephyr/xen/public/domctl.h (hunk 4 above)
Comment (GrygiriiS) [reply include/zephyr/xen/public/domctl.h:23]: It seems enough for this particular exercise. Right change expected to come from https://github.com/xen-troops/zephyr/pull/92 (or similar)

File: boards/arm64/rpi_5/Kconfig.defconfig (hunk 5 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/Kconfig.defconfig]: Huh. Sorry, but it will stay here for RPI5 experiments and exercise demonstrations as I can't even imagine how to resolve DT dependencies between clean RPI5, rpi5 dom0, rpi5 domd (uses at least one hw IP) and xenvm. Lets concider samples/basic/blinky:
 - adding rpi_5.overlay for clean RPI5 will work and might work with dom0 snippet, but will not work for "domD".
 - adding RPI5 specific xenvm.overlay will enable support for RPI5 DomD, bit it can't be done for obvious reasons as xenvm is generic.

File: boards/arm64/rpi_5/Kconfig.board (hunk 2 above)
Comment (firscity) [reply boards/arm64/rpi_5/Kconfig.board]: I am looking on all of these patches from upstream perspective, since I am pretty sure that we will need to upstream it. Unfortunately for you, such approach is not upstreamable.

File: boards/arm64/xenvm/xenvm_defconfig (hunk 1 above)
Comment (firscity) [reply boards/arm64/xenvm/xenvm_defconfig:21]: Oh, I double-checked this config and now I see what is the problem. Okay, let this as is.
We just did not use our Zephyr for building xenvm (since it is not supported/used in any product), only mainline. Will be removed altogether with proper fix.

File: submanifests/zephyr-xenlib.yaml (hunk 3 above)
Comment (firscity) [reply submanifests/zephyr-xenlib.yaml]: You should describe not only what you are adding/doing, but also why and how you want to use it here + same point about upstream perspective.

File: include/zephyr/xen/public/domctl.h (hunk 4 above)
Comment (GrygiriiS) [reply include/zephyr/xen/public/domctl.h:23]: FYI. changes in code not in structs
```
commit bdb1184d4f6bf4e0121fda34a6c1cb51fe270e7d
Author: Jan Beulich <jbeulich@suse.com>
Date:   Thu May 11 13:13:55 2023 +0200

    domctl: bump interface version
    
    The change to XEN_DOMCTL_getdomaininfo was a binary incompatible one,
    and the interface version wasn't bumped yet during the 4.18 release
    cycle.
    
    Fixes: 31c655497461 ("domctl: Modify XEN_DOMCTL_getdomaininfo to fail if domid is not found")
    Signed-off-by: Jan Beulich <jbeulich@suse.com>
    Acked-by: Andrew Cooper <andrew.cooper3@citrix.com>

```

File: boards/arm64/xenvm/xenvm_defconfig (hunk 1 above)
Comment (GrygiriiS) [reply boards/arm64/xenvm/xenvm_defconfig:21]: fixed

File: boards/arm64/rpi_5/Kconfig.board (hunk 2 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/Kconfig.board]: removed

File: submanifests/zephyr-xenlib.yaml (hunk 3 above)
Comment (GrygiriiS) [reply submanifests/zephyr-xenlib.yaml]: dropped

File: boards/arm64/rpi_5/Kconfig.defconfig (hunk 5 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/Kconfig.defconfig]: removed and replaced with snippet

**Comment by GrygiriiS** [comment]
RPI5 commits went upstream, so cherry-picked and noted.
Other comments fixed.

File: drivers/xen/Kconfig [hunk 6]
```diff
@@ -29,6 +29,12 @@ config XEN_GRANT_TABLE_INIT_PRIORITY
 	depends on XEN_GRANT_TABLE
 	default 50
 
+config XEN_EVENTS
```
Comment (firscity) [inline drivers/xen/Kconfig:32]: It will be good to take it to `zephyr-v3.6.0-xt` and upstream this change in separate PRs

File: boards/arm64/xenvm/xenvm.dts [hunk 7]
```diff
@@ -69,7 +69,7 @@
 
 	hypervisor: hypervisor@38000000 {
 		compatible = "xen,xen";
-		reg = <0x00 0x38000000 0x00 0x1000000>;
+		reg = <0x00 0x38000000 0x00 0x10000>;
```
Comment (firscity) [inline boards/arm64/xenvm/xenvm.dts:72]: Should be removed after #96 will be merged

File: boards/arm64/rpi_5/doc/index.rst [hunk 8]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
 
 .. _Raspberry Pi Debug Probe:
    https://www.raspberrypi.com/products/debug-probe/
+
+XEN Dom0
+========
+
+The Raspberry Pi 5 platform can be used to run as Xen Zephyr Dom0. For such purposes the `xen_dom0`
+can be used.
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst]: probably you forgot `snippet`

File: boards/arm64/rpi_5/doc/index.rst [hunk 9]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
...
+
+Run below command as an example of RPI 5 Zephyr build as Dom0:
+
+.. code-block:: bash
+
+   west build -b rpi_5 -p always -S xen_dom0 samples/hello_world
+
+It is expected to be used with special application performing Xen0 functions.
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst]: `Xen0` -> `Xen Domain-0/Dom0`

File: boards/arm64/rpi_5/doc/index.rst [hunk 10]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
...
+
+   The "hypervisor@x" and "memory@x" DT nodes may need to be updated depending on the Xen boot,
+   because normaly Xan will update DT for the target Kernel, but this is not possible in case
+   of Zephyr. See comments in `rpi_5_xen_dom0.dts`.
+
+XEN DomU with HW passthrough
+============================
+
+The Raspberry Pi 5 platform can be used to run as Xen Zephyr DomU with RPI 5 HW support.
+For such purposes the `rpi_5_xen_domd` can be used.
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst]: `snippet`

File: boards/arm64/rpi_5/doc/index.rst [hunk 11]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
...
+
+Run below command as an example of RPI 5 Zephyr build as Dom0:
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst]: `Run below command` -> `Run the command below`

`example of RPI 5 Zephyr build as Dom0` - probably copy-paste issue

File: boards/arm64/rpi_5/doc/index.rst [hunk 12]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
 
 .. _Raspberry Pi Debug Probe:
    https://www.raspberrypi.com/products/debug-probe/
+
+XEN Dom0
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst:163]: Also, I think it is better to make these changes as a part of corresponding commits, which added these functionality (it will be easier to re-arrange it if we will make e.g. DomD snippet as a separate project).

File: boards/arm64/rpi_5/doc/index.rst [hunk 13]
```diff
@@ -159,3 +159,35 @@ When you power on the Raspberry Pi 5, you will see the following output in the s
...
+
+It is expected to be used with special application performing Xen0 functions.
+
+.. note::
+
+   The "hypervisor@x" and "memory@x" DT nodes may need to be updated depending on the Xen boot,
+   because normaly Xan will update DT for the target Kernel, but this is not possible in case
+   of Zephyr. See comments in `rpi_5_xen_dom0.dts`.
+
+XEN DomU with HW passthrough
```
Comment (firscity) [inline boards/arm64/rpi_5/doc/index.rst]: You used "DomD naming" in snippet, so I think it should be the same here.

Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst]: fixed

File: boards/arm64/rpi_5/doc/index.rst (hunk 11 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst]: fixed

File: boards/arm64/rpi_5/doc/index.rst (hunk 10 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst]: fixed

File: boards/arm64/rpi_5/doc/index.rst (hunk 9 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst]: fixed

File: boards/arm64/rpi_5/doc/index.rst (hunk 8 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst]: fixed

File: drivers/xen/Kconfig (hunk 6 above)
Comment (GrygiriiS) [reply drivers/xen/Kconfig:32]: Will note is for integration. and Once we done with basic functionality - I'll revise commits and send those which are generic for integration.

File: boards/arm64/rpi_5/doc/index.rst (hunk 12 above)
Comment (GrygiriiS) [reply boards/arm64/rpi_5/doc/index.rst:163]: I'll keep it in mind.

File: drivers/gpio/gpio_brcmstb.c [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+struct gpio_brcmstb_data {
+	struct gpio_driver_data common;
+
+	DEVICE_MMIO_NAMED_RAM(reg_base);
+	mem_addr_t base;
+};
+
+static int gpio_brcmstb_pin_configure(const struct device *port, gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct gpio_brcmstb_data *data = port->data;
```
Comment (Deedone) [inline drivers/gpio/gpio_brcmstb.c:37]: Null check?

File: drivers/gpio/gpio_brcmstb.c [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+			sys_clear_bit(data->base + GIO_DATA, pin);
+		}
+	}
+
+	return 0;
+}
+
+static int gpio_brcmstb_port_get_raw(const struct device *port, gpio_port_value_t *value)
+{
+	struct gpio_brcmstb_data *data = port->data;
```
Comment (Deedone) [inline drivers/gpio/gpio_brcmstb.c:60]: Same, please check all pointers for NULL value.

File: soc/arm64/bcm2712/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+zephyr_sources_ifdef(CONFIG_ARM_MMU mmu_regions.c)
+
+set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm64/scripts/linker.ld CACHE INTERNAL "")
```
Comment (Deedone) [inline soc/arm64/bcm2712/CMakeLists.txt:4]: Could you add some comments describing why is this needed pls?

File: snippets/rpi_5_xen_domd/README.rst [hunk 17]
```diff
@@ -0,0 +1,17 @@
+.. _rpi_5_xen_domd:
+
+RPI 5 Xen DomD: snippet for XEN HW domain
```
Comment (Deedone) [inline snippets/rpi_5_xen_domd/README.rst:3]: Do we really need separate snippets for DomD and DomUs? AFAIK there is no real difference between them, so what is the reason for explicitly specifying DomD here?

Comment (GrygiriiS) [reply snippets/rpi_5_xen_domd/README.rst:3]: if you look at code ypu'll see DomD based on xenvm DT + passedthrough HW, Dom0 has different Xen specific DT and no HW for now. Those are NOT compatible.

File: soc/arm64/bcm2712/CMakeLists.txt (hunk 16 above)
Comment (GrygiriiS) [reply soc/arm64/bcm2712/CMakeLists.txt:4]: same as above

File: drivers/gpio/gpio_brcmstb.c (hunk 15 above)
Comment (GrygiriiS) [reply drivers/gpio/gpio_brcmstb.c:60]: the same as above

File: drivers/gpio/gpio_brcmstb.c (hunk 14 above)
Comment (GrygiriiS) [reply drivers/gpio/gpio_brcmstb.c:37]: First it's not my patch and it's explicitly marked as back-ported (with "cherry picked from commit 31baddaa51d5b40e6e6394b8be6cb32e1e962f08", so probably right place for comments is in Zephyr mainline https://github.com/zephyrproject-rtos/zephyr/pull/70538.

Second, those functions are not called directly, but through the gpio APIs and null is impossible here.
